core/src/main/scala/kafka/utils/ZkUtils.scala

```diff
@@ -892,3 +871,219 @@ object ZkPath {
+          error("Session has expired while creating %s".format(path))
+        }
+        case _ => {
+          info("ZooKeeper event while creating registration node %s %s".format(path, Code.get(rc)))
+          setResult(Code.get(rc))
+        }
+      }
+    }
+  }
+
+  private class GetDataCallback extends DataCallback {
+      def processResult(rc : Int,
+                 path : String,
+                 ctx : Object,
+                 readData: Array[Byte],
+                 stat : Stat) {
+        Code.get(rc) match {
+          case Code.OK => {
+                  if (checker(ZKStringSerializer.deserialize(readData).asInstanceOf[String], expectedCallerData)) {
+                    info("An ephemeral node [%s] at %s already exists ".format(data, path)
+                      + "it might be because of a past session I have had or connection issues")
+                    checkAndWatch
+                    setResult(Code.OK)
+                } else {
+                    info("An ephemeral node [%s] at %s already exists ".format(data, path)
+                      + "and has been created by someone else")
+                    setResult(Code.NODEEXISTS)
+                }
+          }
+          case Code.NONODE => {
+            info("The ephemeral node [%s] at %s has gone away while reading it, ".format(data, path))
+            setResult(Code.NONODE)
+          }
+          case _ => {
+            setResult(Code.get(rc))
+          }
+        }
+      }
+  }
+
+  private class EphemeralWatcher extends Watcher {
+    def process(event : WatchedEvent) {
+      // if node deleted, then recreate it
+      if(!stop && event.getType == Watcher.Event.EventType.NodeDeleted)
+        createAndWatch
+    }
+  }
+
+  private class ExistsCallback extends StatCallback {
+    def processResult(rc : Int,
+                 path : String,
+                 ctx : Object,
+                 stat : Stat) {
+      Code.get(rc) match {
+        case Code.OK => {}
+        case Code.CONNECTIONLOSS => {
+          // Backoff and try again
+          checkAndWatch
+        }
+        case Code.SESSIONEXPIRED => {
+          error("Session has expired while creating %s".format(path))
+        }
+        case _ => {
+          info("ZooKeeper event while checking if registration node exists %s %s".format(path, Code.get(rc)))
+        }
+      }
+    }
+  }
+
+  private def checkAndWatch() {
+    zkHandle.exists(path,
+                    ephemeralWatcher,
+                    existsCallback,
+                    null)
+  }
+
+  private def createRecursive(prefix : String, suffix : String) {
+    debug("Path: %s, Prefix: %s, Suffix: %s".format(path, prefix, suffix))
+    if(suffix.isEmpty()) {
+      zkHandle.create(prefix,
+                  ZKStringSerializer.serialize(data),
+                  Ids.OPEN_ACL_UNSAFE,
+                  CreateMode.EPHEMERAL,
+                  createCallback,
+                  null)
+    } else {
+      zkHandle.create(prefix,
```

Comment:
For the consumer / server registration case particularly, we tries to handle parent path creation in ZkUtils.makeSurePersistentPathExists, so I feel we should expose the problem that parent path does not exist yet instead trying to hide it in createRecursive.